pkg/controller/stack/stack_controller.go

```diff
@@ -1085,6 +1086,64 @@ func (sess *reconcileStackSession) SetEnvRefsForWorkspace(ctx context.Context, w
 	return nil
 }
 
+func (sess *reconcileStackSession) resolveConfigRefs(ctx context.Context) ([]ConfigKeyValue, error) {
+	allConfigs := make([]ConfigKeyValue, 0)
+	for _, configRefMap := range sess.stack.ConfigRefs {
+		for k, configRef := range configRefMap {
+			// ConfigMap and Literal are special config cases, so they are checked first
+			switch configRef.SelectorType {
+			case shared.ConfigResourceSelectorConfigMap:
+				configMapRef := configRef.ConfigMap
+				if configMapRef != nil {
+					var config corev1.ConfigMap
+					if err := sess.kubeClient.Get(ctx, types.NamespacedName{Name: configMapRef.Name, Namespace: configMapRef.Namespace}, &config); err != nil {
+						return nil, fmt.Errorf("Failed to get the ConfigMap %s on namespace %s: %w", configMapRef.Name, configMapRef.Namespace, err)
+					}
+					// assumes the whole configmaps's data is the config content; try to read as a conventional stack yaml config
+					var configMapContent map[string]any
+					if err := yaml.Unmarshal([]byte(config.Data[configMapRef.Key]), &configMapContent); err != nil {
+						return nil, fmt.Errorf("Failed to read the ConfigMap content as a stack YAML config. Namespace=%s Name=%s: %w", configMapRef.Namespace, configMapRef.Name, err)
+					}
+					structuredConfig := StructuredConfig(configMapContent).Flatten()
+					allConfigs = append(allConfigs, structuredConfig...)
+				}
+			case shared.ConfigResourceSelectorLiteral:
+				literalRef := configRef.Literal
+				if literalRef != nil {
+					// ConfigLiteralRef handles both simple and structured values as json, flattening all keys to build a list of Pulumi key:value configs
+					structuredConfig, err := NewStructuredConfigFromJSON(k, literalRef.Value)
```

Comment:
Already mentioned above, but interpreting literal as an object eliminates the need for `k` while still allowing structure.

```
literal:
  value:
    aws:region: us-west-2
    foo:bar:
      - baz
```